crates/bevy_ecs/src/system/system_registry.rs

```diff
@@ -0,0 +1,299 @@
+use bevy_utils::HashMap;
+use std::any::{Any, TypeId};
+use std::marker::PhantomData;
+
+use crate::system::{Command, IntoSystem, System};
+use crate::world::{Mut, World};
+
+/// A [`System`] that cannot be chained.
+///
+/// [`BoxedSystem`](crate::system::BoxedSystem) is the equivalent type alias for arbitrary `In` and `Out` types.
+pub type UnchainedSystem = Box<dyn System<In = (), Out = ()>>;
+
+/// Stores initialized [`Systems`](crate::system::System), so they can quickly be reused and run
+///
+/// Systems are keyed by their [`TypeId`]: repeated calls with the same function type will reuse cached state,
+/// including for change detection.
+/// Any [`Commands`](crate::system::Commands) generated by these systems (but not other systems), will immediately be applied.
+///
+/// Stored systems cannot be chained: they can neither have an [`In`](crate::system::In) nor return any values.
+#[derive(Default)]
+pub struct SystemRegistry {
+    systems: HashMap<TypeId, UnchainedSystem>,
+}
+
+// User-facing methods
+impl SystemRegistry {
+    /// Registers a system in the [`SystemRegistry`], so then it can be later run.
+    ///
+    /// This only needs to be called manually whenn using [`run_system_by_type_id`](SystemRegistry).
+    #[inline]
+    pub fn register_system<Params, S: IntoSystem<(), (), Params> + 'static>(
+        &mut self,
+        world: &mut World,
+        system: S,
+    ) {
+        let label = system.type_id();
+
+        let mut unchained_system: UnchainedSystem = Box::new(IntoSystem::into_system(system));
+        unchained_system.initialize(world);
+
+        self.systems.insert(label, unchained_system);
+    }
+
+    /// Is the provided `type_id` registered?
+    pub fn type_id_registered(&self, type_id: &TypeId) -> bool {
+        self.systems.contains_key(type_id)
+    }
+
+    /// Runs the system corresponding to the provided [`TypeId`] on the [`World`] a single time
+    pub fn run_system_by_type_id(&mut self, world: &mut World, type_id: TypeId) {
+        let initialized_system = self.systems.get_mut(&type_id).unwrap_or_else(||{panic!{"No system with the `TypeId` {type_id:?} was found. Did you forget to register it?"}});
+        initialized_system.run((), world);
+        initialized_system.apply_buffers(world);
+    }
+
+    /// Runs the supplied system on the [`World`] a single time
+    ///
+    /// If `flush_commands` is true, any [`Commands`](crate::system::Commands) generated will also be applied to the world immediately
+    pub fn run_system<Params, S: IntoSystem<(), (), Params> + 'static>(
+        &mut self,
+        world: &mut World,
+        system: S,
+    ) {
+        let type_id = TypeId::of::<S>();
+        if self.type_id_registered(&type_id) {
+            self.run_system_by_type_id(world, type_id);
+        } else {
+            self.register_system(world, system);
+            self.run_system_by_type_id(world, type_id);
+        }
```

Comment:
Is there a good reason why an unregistered system should be allowed (through implicit registration) to run?

Reply:
Yes: it makes the API dramatically simpler to use.

With implicit registration, the user can call `world.run_system(my_system)` on an ad-hoc basis anywhere in their code, which is fantastic for prototyping and scripting.

Without it, they have to remember to register it, and then remove that registration when the code changes.


Manual registration is *only* there to enable `TypeId` driven workflows. These are much easier to store and work with than an instance of the function, but I would expect manual registration to be hidden in external user-facing code as well :)

Reply:
My original argument against this was that users *could* accidentally run arbitrary systems, which would lead to weird behavior, but the ease of use definitely beats that.
On the second thought, centralizing which functions can be run ad hoc would be helpful for the editor in far future.
Going with that idea, maybe throwing warnings would be a decent middle ground? That way users can still prototype easily, but registration would be enforced.

Reply:
I don't think the warnings are helpful enough to warrant the level of noise. I'm not sure we're protecting against a real class of errors: this isn't any different than just adding systems to a schedule.

Reply:
Yeah but all schedule systems are known at app launch,
from what I understand the one-shot systems can be registered at any point in runtime.
So in editors case it won't be aware of this until it is ran at least once, right?

Reply:
> Yeah but all schedule systems are known at app launch,

This is only true for now. See #279. I suppose we could opt-in to warnings, but IMO that sort of design is much more easily explored once we *have* a way to display all of the systems in the schedule.